### ISSUE TITLE
Increase in memory use is observed when there are multiple calls to wxFileDialog::ShowModal() which might cause application termination

### ISSUE THREAD

**Issue by aperez-hub** [wxWidgets/wxWidgets#23304]
### Description
When we performed saving of bitmap image(overwrite if file exist) repeatedly using wxFileDialog::ShowModal()(to show save dialog) and wxImage::SaveFile(to save image file) while PC memory usage is high, our application was terminated. We are using wxWidget 3.1.5.

To Reproduce:
1. To easily replicate the issue, we increase current memory pressure using the ff command: memory_pressure -l critical
=> In current testing we were able to set Memory Used to 6.5GB/8.0GB based from Activity Monitor
2. Launch application
3. Save image as bitmap (use a filename that already exists)
4. Select Replace when overwrite prompt appears to overwrite existing file
5. Repeat Steps 2 and 3 multiple times

Current Issue: Application will be terminated
Expected Behavior: File will be saved

In console, there was no crash report log. However these following logs are available:
1. System Log
Feb 14 17:59:58 <PCNAME> com.apple.appkit.xpc.openAndSavePanelService[70406](http://192.168.1.181/redmine/issues/5835#fn70406): assertion failed: 19H15: libxpc.dylib + 92759 [3E243A41-030F-38E3-9FD2-7B38C66C35B1]: 0x89
2. XCode Log
2023-02-14 18:01:02.583594+0900 <APPNAME>[70399:3942801] *** Assertion failure in -[NSSavePanel _attachSandboxExtensions:toURL:orURLs:], /AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Sources/AppKit/AppKit-1894.60.100/Nav.subproj/OpenAndSavePanelRemote/NSVBOpenAndSavePanels.m:718

Additional investigations:
1. When PC memory pressure is green (efficient) or NOT high, we CANNOT replicate the issue.
2. When we do multiple save without overwriting(set unique filename every time), we CANNOT replicate the issue. Assumption is there might be more steps/processes performed in overwriting compared to just saving file as is. Since PC memory usage is currently high, accommodating additional processes might result to the issue encountered.
3. com.apple.xpc.openAndSavePanelService process is created the first time Save Dialog was launched.
a. process is retained even after Save Dialog is closed, it is only removed when we close our application
b. After checking behavior of this process in Activity Monitor, what happens is that memory use rises when performing operations related to save, then decreases when those are complete, but never returns to its initial value. Thus, total memory use of this service slowly increases during use. Please see observation below regarding memory used by com.apple.xpc.openAndSavePanelService based from Activity Monitor.
 ``` 
  1. Save Dialog Launched
  => com.apple.xpc.openAndSavePanelService is created, and currently using ~18.5MB 
  2. Save file multiple times
  3. Check com.apple.xpc.openAndSavePanelService memory used
  => increased to 21.4 MB
  4. Do nothing/check if memory will recover
  => still 21.4 MB after ~7minutes
  5. Save file multiple times
  6. Check com.apple.xpc.openAndSavePanelService memory used
  => increased to 21.8 MB
  7. Do nothing/check if memory will recover
  => 22.2 MB after ~3minutes
  8. Do nothing/check if memory will recover
  => 22.3 MB after ~5minutes
  9. Save file multiple times
  => increased to 22.5 MB
  10. Do nothing/check if memory will recover
  => increased to 22.6MB after ~30minutes
  11. Do nothing/check if memory will recover
  => still 22.6 MB after ~60minutes
 ``` 
 c. I think this process is created by NSSavePanel() in function wxFileDialog::ShowWindowModal() in filedlg.mm file. Is this introducing a memory leak issue?

### Platform and version information
1. macOS Catalina 10.15.17
iMac (21.5inch, Late 2013)
Processor 2.7 GHz Quad-Core Intel Core i5
Memory 8 GB 1600 MHz DDR3
Startup Disk MacOS-external-ssd
Graphics Intel Iris Pro 1536 MB
2. wxWidget 3.1.5

This issue is related to:
[https://github.com/wxWidgets/wxWidgets/issues/22173](url)

**Comment by vadz** [comment]
This looks exactly the same as #22173, but I have no idea about what could be causing it.

Could it be a bug in macOS itself? I.e. can you observe the same memory growth when opening the dialogs in the other, non-wx, applications or not?

**Comment by aperez-hub** [comment]
We investigated openAndSavePanelService process memory check using "TextEdit" app of mac pc
--initial save--
00:00:25 ~21.7 MB
--continuous saving--
00:05:50 increase to 25.2 MB
--wait--
00:18:46 start saving again 25.0 MB
--continuous saving--
00:22:32 increase to 25.2 MB
--wait--
00:29:40 25.2 MB
--continuous saving--
00:32:44 increase to 25.6 MB
--wait--
00:43:47 25.4 MB
--continuous saving--
00:45:43 increase to 25.7 MB
--wait--
00:58:16 25.6 MB
--wait--
01:38:07 25.5 MB

Based from above, it has similar patterns with our application.
However, after few hours when I checked the PC, openAndSavePanelService is no longer in Activity Monitor (closed)
Then, when I tried to save again a new openAndSavePanelService process was created again
and it starts with ~22.4 memory (recovered from previous increase).

Additional investigation. Also aside from mac built-in app(textedit), I also tried checking behavior with a macOS app.
I created a macOS app (swift language) that uses NSSavePanel() to save a bitmap image (I used the ColorTrend image from PC App).
Behavior is almost similar with above. As you save, the memory used by openAndSavePanelService increases. However, after leaving it overnight, there is a noticeable decrease in memory used by openAndSavePanelService (28.1 MB -> 23.9 MB)

From the above, the problem does not occur when using the OS API, and it seems to be a problem when using wxFileDialog::ShowModal().
Please investigate if it's a problem with wxFileDialog::ShowModal().

**Comment by vadz** [comment]
@csomor One thing I find suspicious here is that we never release (nor even autorelease) the `NSSavePanel` or `NSOpenPanel` that we create here. Is it done by ARC or do we really leak them?

I.e. could something like this:
```diff
diff --git a/src/osx/cocoa/filedlg.mm b/src/osx/cocoa/filedlg.mm
index a2ba5a6f6f..bce6d255f8 100644
--- a/src/osx/cocoa/filedlg.mm
+++ b/src/osx/cocoa/filedlg.mm
@@ -683,6 +683,7 @@ void wxFileDialog::ModalFinishedCallback(void* panel, int returnCode)
         SendWindowModalDialogEvent ( wxEVT_WINDOW_MODAL_DIALOG_CLOSED  );
     
     [sPanel setAccessoryView:nil];
+    [sPanel release];
 }
 
 #endif // wxUSE_FILEDLG
```
be needed?

**Comment by csomor** [comment]
@vadz no, we don't create it (it's not returned by a call containing Copy or Create), so we don't have to release it, so this cannot be the reason. Actually the panel is a shared instance. 

What is important is that anything added to it is removed (extra controls destroyed). 
It might also be related to energy save / AppNap behavior. 

@aperez-hub is this with a wx sample, or could you please modify a sample, so that it exhibits the same behavior, then I can check which things could be responsible.

**Comment by aperez-hub** [comment]
@vadz , @csomor 
I see. Thank you very much for checking.

**Comment by csomor** [comment]
@aperez-hub I've created PR #23376

**Comment by vadz** [comment]
@aperez-hub Could you please check whether the change in the linked PR fixes the problem for you? It'd be really nice to have a confirmation that it really fixes the problem before merging it. TIA!
